### PR TITLE
[DF] Support complex queries with multiple DISTINCT aggregates

### PR DESCRIPTION
**Pull request by andygrove**
Improves support for `COUNT(DISTINCT expr)` queries:

- Supports GROUP BY
- Supports COUNT DISTINCT on multiple expressions

### CODE REVIEW

File: dask_planner/Cargo.toml [hunk 1]
```diff
@@ -9,19 +9,20 @@ edition = "2021"
...
-tokio = { version = "1.0", features = ["macros", "rt", "rt-multi-thread", "sync", "fs", "parking_lot"] }
-rand = "0.7"
-pyo3 = { version = "0.16.5", features = ["extension-module", "abi3", "abi3-py38"] }
 arrow = { version = "22.0.0", features = ["prettyprint"] }
-datafusion-sql = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-expr = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-common = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-optimizer = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-uuid = { version = "0.8", features = ["v4"] }
+async-trait = "0.1.41"
```
Comment (andygrove) [inline dask_planner/Cargo.toml:13]: I ran `cargo tomlfmt` here and added new dependencies on `log` and `env_logger`

File: dask_planner/src/sql/logical/aggregate.rs [hunk 2]
```diff
@@ -45,47 +45,64 @@ impl PyAggregate {
...
     #[pyo3(name = "getAggregationFuncName")]
     pub fn agg_func_name(&self, expr: PyExpr) -> PyResult<String> {
-        match expr.expr {
-            Expr::AggregateFunction { fun, .. } => Ok(fun.to_string()),
-            Expr::AggregateUDF { fun, .. } => Ok(fun.name.clone()),
-            _ => Err(py_type_err(
-                "Encountered a non Aggregate type in agg_func_name",
-            )),
-        }
+        _agg_func_name(&expr.expr)
```
Comment (andygrove) [inline dask_planner/src/sql/logical/aggregate.rs:48]: The refactoring here is to support aliased aggregate expressions

File: dask_planner/Cargo.toml [hunk 3]
```diff
@@ -9,19 +9,20 @@ edition = "2021"
...
-tokio = { version = "1.0", features = ["macros", "rt", "rt-multi-thread", "sync", "fs", "parking_lot"] }
-rand = "0.7"
-pyo3 = { version = "0.17.1", features = ["extension-module", "abi3", "abi3-py38"] }
 arrow = { version = "22.0.0", features = ["prettyprint"] }
-datafusion-sql = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-expr = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-common = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-datafusion-optimizer = { git="https://github.com/apache/arrow-datafusion/", rev = "12.0.0-rc1" }
-uuid = { version = "0.8", features = ["v4"] }
+async-trait = "0.1.41"
```
Comment (andygrove) [inline dask_planner/Cargo.toml:13]: I ran `cargo tomlfmt` hence the diff here. I also added `log` and `env_logger` dependencies

Comment (charlesbluca) [reply dask_planner/Cargo.toml:13]: Doesn't need to be done in this PR, but would we want a check like this to be run as part of the pre-commit hooks?

File: dask_planner/src/sql/logical/aggregate.rs [hunk 4]
```diff
@@ -45,47 +45,64 @@ impl PyAggregate {
...
+fn _distinct_agg_expr(expr: &Expr) -> PyResult<bool> {
+    match expr {
+        Expr::Alias(expr, _) => _distinct_agg_expr(expr.as_ref()),
+        Expr::AggregateFunction {
+            fun: _,
+            args: _,
+            distinct,
+            filter: _,
+        } => Ok(*distinct),
+        _ => Err(py_type_err(
```
Comment (charlesbluca) [inline dask_planner/src/sql/logical/aggregate.rs:103]: Should we be returning some default value for `distinct` here when we encounter an `AggregateUDF`?

File: tests/integration/test_compatibility.py [hunk 5]
```diff
@@ -276,7 +276,9 @@ def test_join_multi():
     )
 
 
-@pytest.mark.skip(reason="WIP DataFusion")
+@pytest.mark.skip(
+    reason="conflicting aggregation functions: [('count', 'a'), ('count', 'a')]"
+)
```
Comment (charlesbluca) [inline tests/integration/test_compatibility.py:281]: Working on unblocking in #760

File: dask_planner/Cargo.toml (hunk 3 above)
Comment (andygrove) [reply dask_planner/Cargo.toml:13]: Yes, ideally we would add this to the pre commit hook and add checks in CI as well

File: dask_planner/src/sql/logical/aggregate.rs (hunk 4 above)
Comment (andygrove) [reply dask_planner/src/sql/logical/aggregate.rs:103]: That's a good point. I'll take a look.

File: tests/integration/test_compatibility.py [hunk 6]
```diff
@@ -329,7 +348,7 @@ def test_agg_count_distinct_no_group_by():
 
 
 @pytest.mark.skip(
-    reason="WIP DataFusion - https://github.com/dask-contrib/dask-sql/issues/532"
+    reason="conflicting aggregation functions: [('count', 'c'), ('count', 'c')]"
```
Comment (ayushdg) [inline tests/integration/test_compatibility.py:351]: Out of curiosity is this coming from 
```
COUNT(c) AS c_c,
COUNT(DISTINCT c) AS cd_c
```
where the optimization creates a `count(c)` for the latter case as well?

**Pull request merged**